include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -0,0 +1,230 @@
+   * value of `size_t`.
+   *
+   * @return size_t The maximum size of a single allocation supported by this memory resource
+   */
+  size_t get_maximum_allocation_size() const { return std::numeric_limits<size_t>::max(); }
+
+  /**
+   * @brief Allocates memory of size at least `bytes`.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size in bytes of the allocation
+   * @param stream The stream to associate this allocation with
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    if (bytes <= 0) return nullptr;
+
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+    if (stream == cudaStreamDefault || stream == cudaStreamPerThread) {
+      bytes = rmm::detail::align_up(bytes, allocation_alignment);
+      RMM_EXPECTS(
+        bytes <= get_maximum_allocation_size(), rmm::bad_alloc, "Maximum allocation size exceeded");
+
+      if (arena::handles_size(bytes)) { return get_arena()->allocate(bytes); }
+    }
+#endif
+    return upstream_mr_->allocate(bytes, stream);
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * @throws rmm::bad_alloc if `p` is not found
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    if (p == nullptr || bytes <= 0) return;
+
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+    if (stream == cudaStreamDefault || stream == cudaStreamPerThread) {
+      bytes = rmm::detail::align_up(bytes, allocation_alignment);
+      if (arena::handles_size(bytes)) {
+        if (!get_arena()->deallocate(p, bytes, stream)) {
+          deallocate_across_arenas(p, bytes, stream);
+        }
+        return;
+      }
+    }
+#endif
+    upstream_mr_->deallocate(p, bytes, stream);
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p` that was allocated in a different arena.
+   *
+   * @throws rmm::bad_alloc if `p` is not found
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void deallocate_across_arenas(void* p, size_t bytes, cudaStream_t stream)
+  {
+    RMM_CUDA_TRY(cudaStreamSynchronize(stream));
+
+    auto id = std::this_thread::get_id();
+    {
+      lock_guard lock(mtx_);
+      for (auto& kv : arenas_) {
+        if (kv.first != id && kv.second->deallocate(p, bytes)) return;
+      }
+    }
+
+    RMM_FAIL("Allocation not found", rmm::bad_alloc);
```

Comment:
`do_deallocate` can't throw as it's used in dtors which will then end up calling `std::terminate`.

Reply:
Fixed.